drivers/mbox/mbox_ti_omap.c

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2024 Texas Instruments Incorporated.
+ *
+ * TI OMAP Mailbox driver for Zephyr's MBOX model.
+ */
+
+#define DT_DRV_COMPAT ti_omap_mailbox
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/mbox.h>
+#include <zephyr/irq.h>
+#include <zephyr/sys/util_macro.h>
+
+#define LOG_LEVEL CONFIG_MBOX_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ti_omap_mailbox);
+
+/* Helper Macros for MBOX */
+#define DEV_CFG(dev)	((const struct omap_mailbox_config *)((dev)->config))
+#define DEV_DATA(dev)	((struct omap_mailbox_data *)(dev)->data)
+#define DEV_BASE(data)	((struct omap_mailbox_regs *)((data)->mapped_base))
+
+#define MAILBOX_IRQ_NEWMSG(m)		(1 << (2 * (m)))
+#define MAILBOX_IRQ_NOTFULL(m)		(1 << (2 * (m) + 1))
+#define OMAP_MAILBOX_NUM_MSGS		16
+#define MAILBOX_MAX_CHANNELS		16
+#define OMAP_MAILBOX_NUM_USERS		4
+#define MAILBOX_MBOX_SIZE			sizeof(uint32_t)
+
+struct omap_mailbox_data {
+	mbox_callback_t cb[MAILBOX_MAX_CHANNELS];
+	void *user_data[MAILBOX_MAX_CHANNELS];
+	bool channel_enable[MAILBOX_MAX_CHANNELS];
+	uint32_t received_data;
+	void *mapped_base;
+};
+
+struct omap_mailbox_config {
+	uint32_t usr_id;
+};
+
+struct omap_mailbox_irq_regs {
+	uint32_t	status_raw;
+	uint32_t	status_clear;
+	uint32_t	enable_set;
+	uint32_t	enable_clear;
+};
+
+struct omap_mailbox_regs {
+	uint32_t	revision;
+	uint32_t	__pad0[3];
+	uint32_t	sysconfig;
+	uint32_t	__pad1[11];
+	uint32_t	message[OMAP_MAILBOX_NUM_MSGS];
+	uint32_t	fifo_status[OMAP_MAILBOX_NUM_MSGS];
+	uint32_t	msg_status[OMAP_MAILBOX_NUM_MSGS];
+	struct omap_mailbox_irq_regs irq_regs[OMAP_MAILBOX_NUM_USERS];
+};
+
+static void omap_mailbox_isr(const struct device *dev)
+{
+	struct omap_mailbox_data *data = DEV_DATA(dev);
+	const struct omap_mailbox_config *cfg = DEV_CFG(dev);
+
+	volatile struct omap_mailbox_regs *regs = DEV_BASE(data);
+	volatile struct omap_mailbox_irq_regs *irq_regs = &regs->irq_regs[cfg->usr_id];
+	uint32_t flags = irq_regs->status_clear;
```

Comment:
- Is the IRQ status register the same as the IRQ clear register?
- Is the mailbox FIFO based?
If so, there should probably be a loop-until-empty in here somewhere
- A good pattern for ISRs is:
  - read pending ISRs
  - clear pending ISRs
  - process pending ISRs
This prevents race conditions when the ISR is re-triggered during processing (unless they are level sensitive or clearing has side effects)

Reply:
This is the last pending item, yes the mailbox is FIFO based and a loop until empty is needed to clear the FIFO, since most of the testing has been for signalling/single word transfer ,this was not caught , I am working on fixing this.
Ref: https://github.com/torvalds/linux/blob/master/drivers/mailbox/omap-mailbox.c#L206